feat: compute status codes statistics

moreirathomas opened the pull request:
<!-- IMPORTANT: Don't forget to link the issue(s) once the PR created! -->

## Description

Adds stats computation for status codes missing from #6.

~Mirror `timestats` with `codestats`. Incrementally builds the data set at the same location we convert `Report.Records` to a slice of floats for `timestats`~

New specs: simply compute the distribution of status codes within `Records`.

- add `stats.StatusDistribution` and `stats.ComputeStatusDistribution`
- add `benchttp.Record.Code` field and `benchttp.Benchmark.StatusCodes` helper method
- `main` as access to all manner of computing stats and calls them with the correct data slice

## Changes

- `stats.Stats` becomes `stats.Common`: as we said previously, there's nothing _intrinsically_ related to time here, just an interpretation
- `Stats.String` renamed `Common.StringAsTime` to reflect that interpretation possibility (also this is still mainly for logging/debugging logic)

## Notes

`StatusDistribution` is defined flat inside package `stats` because there's no need to extract it inside the worker. As far as worker knows, it is computing stuff around slices of floats/ints it is given. Thus `stats.Common` and `stats.StatusDistribution` are siblings.
Obviously this is not the case on the server.

`stats.Common` is a conscious abstraction choice. As there is really nothing time related, `Common` better suits the abstraction and allows for later similar computation for one request's events and open nicely the myriad of stats available (like `Histogram`).

GregoryAlbouy reviewed (commented):
I'm wondering whether we are interpreting and narrowing the results too soon here. By too soon I mean before storing into the main DB.

If I get it well, in the worker output we lose the status code values: what if, for instance, we want to allow the user to filter its records by status code in the webapp?
What if we change the rules that determine whether a code is a success or a fail? We'd have to update all previous results in the DB by forcing the recompute of their corresponding document.

I may be wrong but I'm not sure this is the ideal place for interpreting the data because it won't handle well the changes 😕 

In any case this definitely needs real specs before diving into it!

c1-ra commented:
I am looking at this while thinking about the database, and maybe we could regroup the status codes into these 5 types:
- informational (100–199)
- successful (200–299)
- redirection (300–399)
- client_error (400–499)
- server_error (500–599)
We would have a little more details than success / fails, but it would still be computed / simplified over handling each status code separately.

moreirathomas committed:
feat: add Code field to Report

- update struct definition
- provide helper to extract the codes
- update firestoreconv package

moreirathomas committed:
misc: change stats.Stats into stats.Common

moreirathomas committed:
feat: add stats.StatusDistribution

c1-ra reviewed (approved):
LGTM

GregoryAlbouy commented on benchttp/benchttp.go:
```diff
@@ -21,7 +21,18 @@ func (b Benchmark) Times() []float64 {
 	return s
 }
 
+// StatusCodes returns the recorded HTTP status codes of the
+// requests as a slice of ints.
+func (b Benchmark) StatusCodes() []int {
+	s := make([]int, len(b.Records))
+	for i, v := range b.Records {
+		s[i] = v.Code
+	}
+	return s
+}
```
[suggestion] Avoid iterating twice on the (potentially very long) records slice:
```suggestion
// StatusCodes returns the recorded HTTP status codes of the
// requests as a slice of ints.
func (b Benchmark) Values() (httpcodes []int, times []float64) {
	httpcodes = make([]int, len(b.Records))
	times = make([]float64, len(b.Records))
	for i, v := range b.Records {
		httpCodes[i] = v.Code
		times[i] = v.Time
	}
	return
}
```
(+ remove `Benchmark.Times)`

moreirathomas committed:
refactor: only loop once on Reports to extract data
